### PR TITLE
Move product canonical url from tpl to controller

### LINKED ISSUE #21152: Missing canonical link on product pages
#### Describe the bug

When you try to override `head.tpl` in child theme, canonical link is missing on product pages.
I'll make a PR very soon.

#### Steps to Reproduce

1. Create `templates/_partials/head.tpl` file in child theme of `classic`
2. Override `{block name='head_seo_title'}...{/block}` (for example)
```
{extends file='parent:_partials/head.tpl'}

{block name='head_seo_title'}[TEST] {$page.meta.title}{/block}
```
3. Canonical link is missing on product pages

#### Additional information

* PrestaShop version: 1.7.x
* PHP version: N/A

### PR DESCRIPTION
**Pull request by micka-fdz**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This PR moves product canonical url on ProductController.
| Type?         | refacto
| Category?     | FO
| BC breaks?    | -
| Deprecations? | -
| Fixed ticket? | Fixes #21152
| How to test?  | Go to product page and see canonical url in `<head>` section.
| Sponsor company | @Creabilis

Original PR: #21154

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/21729)

### CODE REVIEW

File: controllers/front/ProductController.php [hunk 1]
```diff
@@ -83,6 +83,13 @@ public function canonicalRedirection($canonical_url = '')
         }
     }
 
+    public function getCanonicalUrl()
```
Comment (Progi1984) [inline controllers/front/ProductController.php]:
```suggestion
    public function getCanonicalUrl(): string
```

**Review by Progi1984 — COMMENTED** [review]
A small change for improving typing

**Review by NeOMakinG — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

File: controllers/front/ProductController.php [hunk 2]
```diff
@@ -83,6 +83,13 @@ public function canonicalRedirection($canonical_url = '')
         }
     }
 
+    public function getCanonicalUrl(): string
+    {
+        $product = $this->context->smarty->getTemplateVars('product');
```
Comment (PierreRambaud) [inline controllers/front/ProductController.php:89]:
```suggestion
        $product = $this->context->smarty->getTemplateVars('product');
        if (!($product instanceof Product)) {
            return;
        }
```

Make sure the var has the good type. We should never trust and array nor an stdClass.

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
One little thing

**Comment by SD1982** [comment]
HI @micka-fdz it seems there is an error with travis can you please check to fix it ?

**Review by Progi1984 — DISMISSED** [review]

**Review by PierreRambaud — DISMISSED** [review]

**Comment by Progi1984** [comment]
@SD1982 It's good.

**Review by NeOMakinG — DISMISSED** [review]

**Comment by SD1982** [comment]
Hi @micka-fdz I don't see the cannonical url with or without your PR
I have the followings results 
- Classic theme  without PR => canonical url correctly displayed on product page
- Child theme without PR =>  canonical url is not displayed on product page
- Child theme with PR =>  canonical url is not displayed on product page
- Classic theme with PR =>  canonical url is not displayed on product page

I used this zip for the child theme => [childtheme-example.zip](https://github.com/PrestaShop/PrestaShop/files/5678440/childtheme-example.zip)

These results are also confirmed by @khouloudbelguith !!

**Comment by micka-fdz** [comment]
Hi @SD1982, indeed I've just tested again and it doesn't work anymore.

Apparently, this is because of the addition of @PierreRambaud's condition. Where is the mistake?
https://github.com/PrestaShop/PrestaShop/blob/c5606039779ebbda5193922793a3b849ffdcac9a/controllers/front/ProductController.php#L90-L92

**Comment by PierreRambaud** [comment]
> Hi @SD1982, indeed I've just tested again and it doesn't work anymore.
> 
> Apparently, this is because of the addition of @PierreRambaud's condition. Where is the mistake?
> https://github.com/PrestaShop/PrestaShop/blob/c5606039779ebbda5193922793a3b849ffdcac9a/controllers/front/ProductController.php#L90-L92

Yes the type is not Product but ProductLazyArray, I fixed it.

**Review by PierreRambaud — APPROVED** [review]

**Comment by PierreRambaud** [comment]
And also fix the method name which is `getCanonicalURL` instead of `getCanonicalUrl`

**Review by PululuK — APPROVED** [review]

**Comment by SD1982** [comment]
LGTM now Thanks @micka-fdz !!

**Comment by Progi1984** [comment]
Thanks @micka-fdz & @SD1982

**Pull request merged**